### PR TITLE
Incremental solvable orders cache update

### LINKED ISSUE #2831: chore: Improve time to create Auction
### Background
Updating solvable orders (ie creating a new auction) currently takes >2s with some pretty heavy outliers ([logs](https://production-6de61f.kb.eu-central-1.aws.cloud.es.io/app/r/s/ALsEK))

This makes it hard to bring CoW protocol's auction rate down to one batch per block as simply creating up to date state would take >15% of the total time we have at hand. We should at least be able to half this time (if not getting it down even more)

### Details
The first step should be to get a better understanding on which of the different parts in the auction creation takes the most time.

It's very likely that this is related to us computing the auction always "from scratch". We should see if we can switch to a more incremental way where only the necessary "on chain state" is queried after the block for the next auction has been processed, but all other things are kept ready and warm throughout the 12s (e.g. whatever we do when a new order is placed should happen as soon as we notice a new order)

### Acceptance criteria
p95 time to create a new auction is <1s

### PR DESCRIPTION
**Pull request by squadgazzz**
# Description
> Updating solvable orders (ie creating a new auction) currently takes >2s with some pretty heavy outliers ([logs](https://production-6de61f.kb.eu-central-1.aws.cloud.es.io/app/r/s/ALsEK))
> 
> This makes it hard to bring CoW protocol's auction rate down to one batch per block as simply creating up to date state would take >15% of the total time we have at hand. We should at least be able to half this time (if not getting it down even more)

In order to relieve the situation, it was proposed to introduce incremental solvable orders cache update, which selects all the solvable orders using the old heavy query only at startup, stores the latest received order's creation timestamp in memory, and then makes much faster incremental bounded queries to the orders and additional tables that select fewer data and executes faster.

# Changes

Since incremental fetching retrieves orders created/cancelled after the specific timestamps, it is also required now to fetch orders that have any onchain update based on the last fetched block number. Having said that, the data needs to be fetched within a single TX, so there is no way to run all the queries in parallel.

1. If the current solvable orders cache is empty, execute the original heavy SQL query to fetch all current solvable orders and store them in memory.
2. Otherwise, fetch full orders that created or cancelled after the last stored timestamp and also find UIDs of the order that have any onchain data updated after the latest observed block number. This includes fetching data from the following tables: trades, ethflow_data, order_execution, invalidations, onchain_order_invalidations, onchain_placed_orders, presignature_events.
3. Fetch quotes for all the collected orders.
4. Add all the newly received orders to the cache.
5. Filter out all the orders that are one of: contain on-chain errors, expired, fulfilled, invalidated.
6. Calculate the latest observed order creation timestamp.
7. Continue with the regular auction creation process.

As a result, we now have 3 SQL queries where each executes in ~50ms instead of a single one taking ~2s.

## How to test
New DB tests. Existing e2e tests.

## Related Issues

Fixes #2831

### CODE REVIEW

File: crates/autopilot/src/solvable_orders.rs [hunk 1]
```diff
@@ -156,8 +158,95 @@ impl SolvableOrdersCache {
...
+        new_quotes: HashMap<domain::OrderUid, domain::Quote>,
+    ) -> Result<boundary::SolvableOrders> {
+        let now = now_in_epoch_seconds();
+        let mut orders = current_orders.orders.clone();
+        // Remove expired orders.
+        orders.retain(|_uid, order| order.data.valid_to >= now);
+        let mut quotes = current_orders.quotes.clone();
+        let mut latest_trade_block = current_orders.latest_settlement_block;
+
+        for new_order in new_orders {
```
Comment (MartinquaXD) [inline crates/autopilot/src/solvable_orders.rs:183]: This logic seems pretty hard to grasp.
Could we first try to implement a simpler approach in 2 phases?
1. blindly add new orders and update existing orders based on trades
2. go over all orders and filter out all the expired, filled, etc. ones

Since caching is known for being tricky I would only consider more optimal implementations if we definitely need the extra performance.

Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs:190]: IMO, that looks even more complicated. Since we don't have trades data for new orders, the process will look as follows:
1. Create new full orders based on the trades data and insert them into the cache.
2. Iterate over the fetched orders again to find and remove expired/fulfilled/canceled orders from the cache.
3. Iterate over the cache to do the same with the old existing orders.

Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs:190]: IMO, that looks even more complicated. Since we don't have trades data for new orders, the process will look as follows:

1. Create new full orders based on the trades data and insert them into the cache.
2. Iterate over the fetched orders(that don't have associated trades data) again to find and remove expired/fulfilled/canceled orders from the cache.
3. Iterate over the cache to do the same with the old existing orders.

Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs:190]: > Also why do we need to have trades separately? 

I was thinking about this today and something stopped me from doing this. But after reverting to that one more time, I don't really see any blocker.

Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs]: Damn, failing the test reminded me what the issue was initially. Recent trades might contain data for orders created before the latest timestamp. Will revert the latest changes...

Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs]: It turned out that not only trade data need to be fetched separately. Update the PR description.

Comment (MartinquaXD) [reply crates/autopilot/src/solvable_orders.rs]: > Having said that, the data needs to be fetched within a single TX, so there is no way to run all the queries in parallel.

> Recent trades might contain data for orders created before the latest timestamp.

Those 2 issues could still be resolved by the simpler approach, no?
I was imagining a single (probably somewhat complex) query that just returns the list of uids where **something** changed since last time. This can be parallelized as postgres sees fit.
Afterwards use the eixisting query to retrieve "full orders" from the DB for each uid, add/replace them in the cache and finally filter out all the orders that are not desirable.

Would this not make things easier?

Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs]: Found another potential corner case.
Since [onchain events are split into chunks](https://github.com/cowprotocol/services/blob/e7de6bfd0526e00b55fda817b33a0fde715c8a61/crates/shared/src/event_handling.rs#L283) with hardcoded value, there is a possibility for a case when a data for one block is separated into different chunks and order updates fetched in between these two operations making the logic store block X as the last seen, but in fact, there are remaining records.

File: crates/autopilot/src/database/auction.rs [hunk 2]
```diff
@@ -73,24 +73,18 @@ impl Postgres {
...
             .await?;
-        let orders: Vec<Order> = database::orders::solvable_orders(&mut ex, min_valid_to as i64)
-            .map(|result| match result {
-                Ok(order) => full_order_into_model_order(order),
-                Err(err) => Err(anyhow::Error::from(err)),
-            })
-            .try_collect()
-            .await?;
+        let orders: HashMap<domain::OrderUid, Order> =
+            database::orders::solvable_orders(&mut ex, min_valid_to as i64)
```
Comment (m-lord-renkse) [inline crates/autopilot/src/database/auction.rs:77]: Please, don't cast an `u32` to an `i64` like that, use better `try_from()`. See https://github.com/cowprotocol/services/pull/2828#discussion_r1689617078

File: crates/autopilot/src/database/auction.rs [hunk 3]
```diff
@@ -73,24 +73,18 @@ impl Postgres {
...
-        let quotes = self
-            .read_quotes(
-                orders
-                    .iter()
-                    .map(|order| domain::OrderUid(order.metadata.uid.0))
-                    .collect::<Vec<_>>()
-                    .iter(),
-            )
-            .await?;
+        let quotes = self.read_quotes(orders.keys()).await?;
```
Comment (m-lord-renkse) [inline crates/autopilot/src/database/auction.rs:87]: Couldn't this `await` be parallelized with the `latest_settlement_block()` one? :thinking:

File: crates/autopilot/src/infra/persistence/mod.rs [hunk 4]
```diff
@@ -382,6 +390,157 @@ impl Persistence {
         Ok(solution)
     }
 
+    pub async fn solvable_order_after(
```
Comment (m-lord-renkse) [inline crates/autopilot/src/infra/persistence/mod.rs:393]: Please, can you extend this function with a comment explaining what exactly it does? :pray:

File: crates/autopilot/src/infra/persistence/mod.rs [hunk 5]
```diff
@@ -382,6 +390,157 @@ impl Persistence {
...
+                })
+                .try_collect()
+                .await?
+        };
+
+        let next_order_uids = next_orders.keys().cloned().collect::<HashSet<_>>();
+
+        // Interactions table doesn't contain block number, so we need to update
+        // interactions for orders from the cache.
+        let current_order_interactions = {
```
Comment (m-lord-renkse) [inline crates/autopilot/src/infra/persistence/mod.rs:444]: `current_order_interactions` and `updated_quotes` could be parallelized.

File: crates/autopilot/src/infra/persistence/mod.rs [hunk 6]
```diff
@@ -382,6 +390,157 @@ impl Persistence {
...
+            latest_settlement_block,
+            min_valid_to,
+        )
+    }
+
+    fn build_solvable_orders(
+        current_orders: &boundary::SolvableOrders,
+        current_orders_interactions: HashMap<domain::OrderUid, model::order::Interactions>,
+        next_orders: HashMap<domain::OrderUid, model::order::Order>,
+        mut new_quotes: HashMap<domain::OrderUid, domain::Quote>,
```
Comment (m-lord-renkse) [inline crates/autopilot/src/infra/persistence/mod.rs:485]: for consistency, shouldn't it be `next_quotes`? :thinking:

File: crates/autopilot/src/infra/persistence/mod.rs [hunk 7]
```diff
@@ -382,6 +390,157 @@ impl Persistence {
...
+        latest_settlement_block: u64,
+        min_valid_to: u32,
+    ) -> anyhow::Result<boundary::SolvableOrders> {
+        let mut orders = current_orders.orders.clone();
```
Comment (m-lord-renkse) [inline crates/autopilot/src/infra/persistence/mod.rs:489]: why not to pass the `current_orders` as mutable directly? so we avoid an internal clone

File: crates/autopilot/src/infra/persistence/mod.rs [hunk 8]
```diff
@@ -382,6 +390,157 @@ impl Persistence {
...
+        }
+
+        // Filter out all the invalid orders.
+        orders.retain(|_uid, order| {
+            let expired = order.data.valid_to < min_valid_to
+                || order
+                    .metadata
+                    .ethflow_data
+                    .as_ref()
+                    .is_some_and(|data| data.user_valid_to < min_valid_to as i64);
```
Comment (m-lord-renkse) [inline crates/autopilot/src/infra/persistence/mod.rs:510]: ditto

File: crates/autopilot/src/solvable_orders.rs [hunk 9]
```diff
@@ -170,29 +174,62 @@ impl SolvableOrdersCache {
...
+        // (`persistence::solvable_orders`). This prevents situations where
+        // starting the service with a large existing DB would cause
+        // the incremental query to load all unfiltered orders into memory, potentially
+        // leading to OOM issues.
+        let (db_solvable_orders, previous_creation_timestamp) = {
+            let lock = self.cache.lock().await;
+            match &*lock {
+                Some(cache) if cache.last_order_creation_timestamp > DateTime::<Utc>::MIN_UTC => (
+                    self.persistence
+                        .solvable_order_after(
```
Comment (m-lord-renkse) [inline crates/autopilot/src/solvable_orders.rs:192]: Here we are locking the cache for the duration of the function `solvable_order_after`

File: crates/autopilot/src/database/auction.rs (hunk 2 above)
Comment (squadgazzz) [reply crates/autopilot/src/database/auction.rs:77]: Why? u32 can be safely casted to i64, can't it?

File: crates/autopilot/src/database/auction.rs (hunk 3 above)
Comment (squadgazzz) [reply crates/autopilot/src/database/auction.rs:87]: No, since there is an opened transaction.

File: crates/autopilot/src/infra/persistence/mod.rs (hunk 5 above)
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/mod.rs:444]: ditto: https://github.com/cowprotocol/services/pull/2923#discussion_r1740655235

File: crates/autopilot/src/database/auction.rs (hunk 2 above)
Comment (MartinquaXD) [reply crates/autopilot/src/database/auction.rs:77]: Could even be an `.into()` since any `u32` fits into `i64`.

Comment (m-lord-renkse) [reply crates/autopilot/src/database/auction.rs:77]: oh yep, you are right, it is `u32` not `u64` :ok_hand: nevertheless, I would then use `from` for the reason that if in the future the type changes, it doesn't create problems since the `from()` would complain, but the `as` wouldn't (I do not like `as` casting for this precise reason).

File: crates/autopilot/src/infra/persistence/mod.rs (hunk 7 above)
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/mod.rs:489]: Your suggestion could leave the cache in an unpredictable state in case something fails in the middle.

File: crates/autopilot/src/solvable_orders.rs [hunk 10]
```diff
@@ -170,29 +174,62 @@ impl SolvableOrdersCache {
...
+                ),
+            }
+        };
+
+        let latest_creation_timestamp = db_solvable_orders
+            .orders
+            .values()
+            .map(|order| order.metadata.creation_date)
+            .max()
+            .unwrap_or(previous_creation_timestamp);
```
Comment (MartinquaXD) [inline crates/autopilot/src/solvable_orders.rs:228]: Does this actually reduce the latest timestamp if the last created order gets cancelled?
To get around that you could chain `[previous_creation_timestamp]` onto the iterator and always `.uwrap()`.
Or you do `std::cmp::max(max(order.metadata.creation_data),  previous_creation_timestamp)`

File: crates/autopilot/src/solvable_orders.rs [hunk 11]
```diff
@@ -170,29 +174,62 @@ impl SolvableOrdersCache {
...
+        // (`persistence::solvable_orders`) returned some orders. Until then, `MIN_UTC`
+        // is used to indicate that no orders have been found yet by
+        // (`persistence::solvable_orders`). This prevents situations where
+        // starting the service with a large existing DB would cause
+        // the incremental query to load all unfiltered orders into memory, potentially
+        // leading to OOM issues.
+        let (db_solvable_orders, previous_creation_timestamp) = {
+            let lock = self.cache.lock().await;
+            match &*lock {
+                Some(cache) if cache.last_order_creation_timestamp > DateTime::<Utc>::MIN_UTC => (
```
Comment (MartinquaXD) [inline crates/autopilot/src/solvable_orders.rs]: Could make sense to have a named constant for the `MIN_UTC` sentinel value.

File: crates/autopilot/src/infra/persistence/mod.rs [hunk 12]
```diff
@@ -382,6 +390,157 @@ impl Persistence {
...
+        current_orders: &boundary::SolvableOrders,
+        current_orders_interactions: HashMap<domain::OrderUid, model::order::Interactions>,
+        next_orders: HashMap<domain::OrderUid, model::order::Order>,
+        mut new_quotes: HashMap<domain::OrderUid, domain::Quote>,
+        latest_settlement_block: u64,
+        min_valid_to: u32,
+    ) -> anyhow::Result<boundary::SolvableOrders> {
+        let mut orders = current_orders.orders.clone();
+
+        // Update order interactions.
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/mod.rs:488]: I was worried about this logic: https://github.com/cowprotocol/services/blob/2cc91259c1925e49a9df63e0cdef3ee584a333de/crates/autopilot/src/database/onchain_order_events/ethflow_events.rs#L84-L107, which overrides the data.

File: crates/database/src/orders.rs [hunk 13]
```diff
@@ -687,6 +688,25 @@ pub fn solvable_orders(
...
+    ex: &'a mut PgConnection,
+    uids: &'a [OrderUid],
+    after_timestamp: DateTime<Utc>,
+) -> BoxStream<'a, Result<FullOrder, sqlx::Error>> {
+    #[rustfmt::skip]
+const OPEN_ORDERS_AFTER: &str = const_format::concatcp!(
+    "SELECT ", ORDERS_SELECT,
+    " FROM ", ORDERS_FROM,
+    " LEFT OUTER JOIN ethflow_orders eth_o on eth_o.uid = o.uid ",
+    " WHERE (o.creation_timestamp > $1 OR o.cancellation_timestamp > $1 OR o.uid = ANY($2))",
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/database/src/orders.rs:701]: This query returns newly created/cancelled orders and also fetches all the updates from other tables for orders created before, but updated since the last fetch. Didn't really get how this should be changed.

File: crates/autopilot/src/infra/persistence/mod.rs (hunk 7 above)
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/mod.rs:492]: I mean, if the orders in the cache were updated, but not the auction.

File: crates/autopilot/src/solvable_orders.rs (hunk 10 above)
Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs]: Good catch, fixed.

File: crates/autopilot/src/infra/persistence/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/autopilot/src/infra/persistence/mod.rs:494]: Hmm, interesting. I believe updating the interactions is also not necessary there.
Interactions are tied to the signed `appdata` hash of the order. But that hash also influences the order uid so unless you have an exceedingly unlikely hash collision changing the interactions will result in a completely different order uid making this overwrite operation pointless. 🤔

File: crates/database/src/orders.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/database/src/orders.rs:701]: > So I guess what's confusing me here is that uid could easily contain an order which was created way before after_timestamp and the function would still return data from that order.

Maybe this can be solved by renaming the function instead of introducing one more SQL query?

[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/database/src/orders.rs:701]: > 1. fetch all uids of orders created or cancelled after the timestamp
> 2. add those to the list of previously open uids

Or do you mean here to replace with it the SQL query that finds updated UIDs? In that case if we have ~2k orders for each auction, the result will contain x2(~4k) orders. By searching only for the updated UIDs, that amount is expected to decrease.

File: crates/autopilot/src/database/auction.rs (hunk 3 above)
Comment (m-lord-renkse) [reply crates/autopilot/src/database/auction.rs:90]: I don't understand. Are you saying that in order to get the quote, we need to first read the latest settlement block?

Comment (squadgazzz) [reply crates/autopilot/src/database/auction.rs:90]: There is an opened tx `https://github.com/cowprotocol/services/blob/a10b4e33c7c8ed10e65d8f1348a9143ad1d80188/crates/autopilot/src/infra/persistence/mod.rs#L400-L406`. We should fetch all the data using a single DB snapshot even for quotes since there is a logic of updating them based on the onchain data: https://github.com/cowprotocol/services/blob/e7de6bfd0526e00b55fda817b33a0fde715c8a61/crates/autopilot/src/database/onchain_order_events.rs#L329-L342

Comment (m-lord-renkse) [reply crates/autopilot/src/database/auction.rs:90]: In that case, shouldn't we use the same `tx` for fetching the quote?

File: crates/autopilot/src/infra/persistence/mod.rs [hunk 14]
```diff
@@ -382,6 +387,129 @@ impl Persistence {
...
+            .await?;
+
+        // Find order uids for orders that were updated after the given block.
+        let updated_order_uids: Vec<database::OrderUid> = {
+            let _timer = Metrics::get()
+                .database_queries
+                .with_label_values(&["updated_order_uids"])
+                .start_timer();
+
+            database::orders::updated_order_uids_after(&mut tx, after_block)
```
Comment (m-lord-renkse) [inline crates/autopilot/src/infra/persistence/mod.rs:415]: Coming to think about it, cannot we do this query and the next ones in a single roundtrip? or at least to minimize it? we have 3/4 roundtrips to the database.

File: crates/autopilot/src/database/auction.rs (hunk 3 above)
Comment (squadgazzz) [reply crates/autopilot/src/database/auction.rs:90]: Oh, right. Looks like I reverted the wrong change during one of the reworks. Thanks!

File: crates/autopilot/src/infra/persistence/mod.rs (hunk 14 above)
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/mod.rs:416]: Please elaborate. Not sure which queries could be combined.

File: crates/autopilot/src/database/onchain_order_events/ethflow_events.rs [hunk 15]
```diff
@@ -124,6 +124,7 @@ impl OnchainOrderParsing<EthFlowData, EthFlowDataForDb> for EthFlowOnchainOrderP
             // in order to wrap all existing ether to weth, such that the eth can be used as
             // WETH by the cow protocol
             pre_interaction: Interaction {
+                order_uid: order.uid,
```
Comment (MartinquaXD) [inline crates/autopilot/src/database/onchain_order_events/ethflow_events.rs:127]: Not sure why we suddenly need the uid for the interactions. 🤔

File: crates/autopilot/src/infra/persistence/mod.rs [hunk 16]
```diff
@@ -382,6 +388,156 @@ impl Persistence {
...
+                Ok(order) => full_order_into_model_order(order)
+                    .map(|order| (domain::OrderUid(order.metadata.uid.0), order)),
+                Err(err) => Err(anyhow::Error::from(err)),
+            })
+            .try_collect()
+            .await?
+        };
+
+        // Fetch quotes for new orders and also update them for the cached ones since
+        // they could also be updated.
```
Comment (MartinquaXD) [inline crates/autopilot/src/infra/persistence/mod.rs:445]: This is also needed because ethflow orders could theoretically be reorged?
We should really find a way to implement ethflow orders in a nicer way to remove all those ugly edge cases. :/

File: crates/autopilot/src/solvable_orders.rs [hunk 17]
```diff
@@ -168,31 +172,84 @@ impl SolvableOrdersCache {
     /// the case in unit tests, then concurrent calls might overwrite each
     /// other's results.
     pub async fn update(&self, block: u64) -> Result<()> {
+        const INITIAL_ORDER_CREATION_TIMESTAMP: DateTime<Utc> = DateTime::<Utc>::MIN_UTC;
```
Comment (MartinquaXD) [inline crates/autopilot/src/solvable_orders.rs:175]: After thinking more about it do we even need this sentinel value?
The cached order can already be optional and I would expect once we populate the cache there also has to be a non-zero timestamp in there so we can determine by the cache being `None` that we need to fetch ALL open orders instead of the incremental update.

File: crates/autopilot/src/solvable_orders.rs [hunk 18]
```diff
@@ -168,31 +172,84 @@ impl SolvableOrdersCache {
...
+            .map(|order| order.metadata.creation_date)
+            .max()
+            .map_or(previous_creation_timestamp, |max_creation_timestamp| {
+                std::cmp::max(max_creation_timestamp, previous_creation_timestamp)
+            });
+        let orders = db_solvable_orders
+            .orders
+            .values()
+            .cloned()
+            .collect::<Vec<_>>();
```
Comment (MartinquaXD) [inline crates/autopilot/src/solvable_orders.rs:238]: The new code until here could be moved into a separate function.

File: crates/autopilot/src/infra/persistence/mod.rs (hunk 16 above)
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/mod.rs:445]: > This is also needed because ethflow orders could theoretically be reorged?

Because of this code:
https://github.com/cowprotocol/services/blob/e7de6bfd0526e00b55fda817b33a0fde715c8a61/crates/autopilot/src/database/onchain_order_events.rs#L329-L342

File: crates/autopilot/src/solvable_orders.rs (hunk 17 above)
Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs:175]: With `None` some of the e2e tests fail. It's probably worth a separate PR to fix it.

Comment (MartinquaXD) [reply crates/autopilot/src/solvable_orders.rs:175]: Let's please understand at least what causes the tests to fail. I don't want us to check in code without understanding why it's actually needed.

File: crates/autopilot/src/infra/persistence/mod.rs (hunk 14 above)
Comment (m-lord-renkse) [reply crates/autopilot/src/infra/persistence/mod.rs:416]: sure! so, my concern is that we do these two queries to the orders database:

- `updated_order_uids_after`
- `open_orders_by_time_and_uids`

The thing is that the output of `updated_order_uids_after` is used as an input for `open_orders_by_time_and_uids`. So, we potentially could combine those queries into one roundtrip (of course, the query would be relatively complex), but we do have already such complexity levels in queries (e.g. see `ORDERS_SELECT`).

File: crates/autopilot/src/solvable_orders.rs [hunk 19]
```diff
@@ -337,6 +340,71 @@ impl SolvableOrdersCache {
...
+                            cache.last_order_creation_timestamp,
+                            cache.solvable_orders.latest_settlement_block,
+                        ))
+                    }
+                    _ => None,
+                }
+            };
+
+            let min_valid_to =
+                now_in_epoch_seconds() + self.min_order_validity_period.as_secs() as u32;
```
Comment (m-lord-renkse) [inline crates/autopilot/src/solvable_orders.rs:376]: Extremely dangerous casting

File: crates/autopilot/src/solvable_orders.rs [hunk 20]
```diff
@@ -337,6 +340,71 @@ impl SolvableOrdersCache {
             .collect()
     }
 
+    /// Returns current solvable orders along with the latest order creation
+    /// timestamp.
+    async fn get_solvable_orders(&self) -> Result<(SolvableOrders, DateTime<Utc>)> {
```
Comment (m-lord-renkse) [inline crates/autopilot/src/solvable_orders.rs:345]: Getting the latest order creation timestamp can be done with the `SolvableOrders`. I believe it shouldn't be the responsibility of this function.

File: crates/database/src/orders.rs [hunk 21]
```diff
@@ -1532,7 +1583,63 @@ mod tests {
             .await
             .unwrap();
 
-        assert!(get_order(&mut db, 2).await.is_none());
+        assert!(get_full_order(&mut db, 2).await.is_none());
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn postgres_open_orders_by_time_and_uids() {
```
Comment (m-lord-renkse) [inline crates/database/src/orders.rs:1591]: Please, could you add comments within the test to explain each step :pray:

File: crates/autopilot/src/solvable_orders.rs (hunk 17 above)
Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs]: That requires refactoring. Some of the code depends on auction always present in the DB, which is used across many of the e2e tests.
https://github.com/cowprotocol/services/blob/fb8dad32106a2a42bf9cfb5e6a7db60755fa69f3/crates/e2e/src/setup/services.rs#L303-L317
Also:
https://github.com/cowprotocol/services/blob/db3a1b578c1caddd8ec428a98db5a8039177e2cb/crates/e2e/src/setup/services.rs#L304-L317

File: crates/autopilot/src/solvable_orders.rs (hunk 20 above)
Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs:345]: This function has to return either `previous_creation_timestamp` or `latest_creation_timestamp`. Since the former is not used anywhere else, I decided to keep it in this function.

File: crates/autopilot/src/infra/persistence/mod.rs (hunk 14 above)
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/mod.rs]: Nice idea 👍  Unified queries.

File: crates/autopilot/src/solvable_orders.rs (hunk 17 above)
Comment (MartinquaXD) [reply crates/autopilot/src/solvable_orders.rs]: Would be great to refactor since the sentinel value really doesn't seem necessary but IMO that can happen after this gets merged so we can test this a bit on staging before merging to prod on Tuesday.

File: crates/autopilot/src/solvable_orders.rs [hunk 22]
```diff
@@ -337,6 +340,72 @@ impl SolvableOrdersCache {
...
```
Comment (MartinquaXD) [inline crates/autopilot/src/solvable_orders.rs:345]: Is it possible to simplify this query a lot by always using `solvable_orders_after()`?
All you need for that is the list of known orders, latest timestamp and latest block.
If the cache is populated you just take those values and if it's not you should be able to just use defaults for everything, no?

**Review by MartinquaXD — COMMENTED** [review]
Would like to simplify the code inside `SolvableOrdersCache` further if possible but if my suggestion doesn't work I'm fine with merging now and refactoring later to already get confidence in the change by running it in staging.

File: crates/autopilot/src/solvable_orders.rs (hunk 22 above)
Comment (squadgazzz) [reply crates/autopilot/src/solvable_orders.rs:345]: > Is it possible to simplify this query a lot by always using solvable_orders_after()?

`solvable_orders_after` doesn't filter orders in the SQL query, so it would return the whole table at the first start if the creation timestamp/block number contains default values. I tried to explain this in the comment:
https://github.com/cowprotocol/services/blob/1447bba2fe04c1f87d8eddd1819e8ac105eda295/crates/autopilot/src/solvable_orders.rs#L348-L358

**Review by MartinquaXD — APPROVED** [review]
Alright, let's get this merged to make strides on the roadmap but let's try to keep simplifying the code further.
